pkg/api/validation/validation.go

```diff
@@ -1398,3 +1400,90 @@ func ValidateEndpointsUpdate(oldEndpoints, newEndpoints *api.Endpoints) errs.Val
 	allErrs = append(allErrs, validateEndpointSubsets(newEndpoints.Subsets).Prefix("subsets")...)
 	return allErrs
 }
+
+// ValidateSecurityContext tests to make sure that a security context is set and that the settings
+// in the security context do not conflict with settings on the container.
+func ValidateSecurityContext(container *api.Container) errs.ValidationErrorList {
+	allErrs := errs.ValidationErrorList{}
+	mergeSecurityContextAndContainer(container)
+	sc := container.SecurityContext
+	if container.Privileged != *sc.Privileged {
+		allErrs = append(allErrs, errs.NewFieldInvalid("privileged", *sc.Privileged, "privileged conflicts with container.privileged"))
+	}
+	if !reflect.DeepEqual(sc.Capabilities.Add, container.Capabilities.Add) {
+		allErrs = append(allErrs, errs.NewFieldInvalid("capabilities.add", sc.Capabilities.Add, "capabilities.add conflicts with container.capabilities.add"))
+	}
+	if !reflect.DeepEqual(sc.Capabilities.Drop, container.Capabilities.Drop) {
+		allErrs = append(allErrs, errs.NewFieldInvalid("capabilities.drop", sc.Capabilities.Drop, "capabilities.drop conflicts with container.capabilities.drop"))
+	}
+	if sc.SELinuxOptions != nil {
+		allErrs = append(allErrs, validateSELinuxOptions(sc.SELinuxOptions).Prefix("seLinuxOptions")...)
+	}
+	if sc.RunAsUser != nil {
+		if *sc.RunAsUser < 0 {
+			allErrs = append(allErrs, errs.NewFieldInvalid("runAsUser", *sc.RunAsUser, "runAsUser cannot be negative"))
+		}
+	}
+	return allErrs
+}
+
+// validateSELinux ensures that the disable and label options are not mixed
+func validateSELinuxOptions(seLinux *api.SELinuxOptions) errs.ValidationErrorList {
+	allErrs := errs.ValidationErrorList{}
+	if seLinux.Disabled {
+		if len(seLinux.Level) > 0 || len(seLinux.Role) > 0 || len(seLinux.Type) > 0 || len(seLinux.User) > 0 {
+			allErrs = append(allErrs, errs.NewFieldInvalid("disabled", seLinux.Disabled, "cannot set disabled and labels"))
+		}
+	}
+	return allErrs
+}
+
+// mergeSecurityContextAndContainer performs a merge of the settings that exist on both the container
+// and the securitycontext.  This includes the Privileged field and Capabilities field
+// merge rules are:
+// 1.  if the container has no security context then create one and use all settings from the container
+// 2.  if the container has a security context but no capabilities defined then create one and use
+//		all capability settings from the container
+// 3.  if the security context has capabilities defined merge upwards or downwards by merging a non-zero
```

Comment:
Probably - can't think of a user accidentally getting into this state.  Since validation is probably moving to external it seems reasonable.

> On Apr 28, 2015, at 7:37 PM, Eric Tune <notifications@github.com> wrote:
> 
> In pkg/api/validation/validation.go:
> 
> > +	allErrs := errs.ValidationErrorList{}
> > +	if seLinux.Disabled {
> > +		if len(seLinux.Level) > 0 || len(seLinux.Role) > 0 || len(seLinux.Type) > 0 || len(seLinux.User) > 0 {
> > +			allErrs = append(allErrs, errs.NewFieldInvalid("disabled", seLinux.Disabled, "cannot set disabled and labels"))
> > +		}
> > +	}
> > +	return allErrs
> > +}
> > +
> > +// mergeSecurityContextAndContainer performs a merge of the settings that exist on both the container
> > +// and the securitycontext.  This includes the Privileged field and Capabilities field
> > +// merge rules are:
> > +// 1.  if the container has no security context then create one and use all settings from the container
> > +// 2.  if the container has a security context but no capabilities defined then create one and use
> > +//		all capability settings from the container
> > +// 3.  if the security context has capabilities defined merge upwards or downwards by merging a non-zero
> can we just reject if it has nonempty nonequal capabilities structs?
> 
> —
> Reply to this email directly or view it on GitHub.
>